File: server/controllers/goodsController.js
/**
 * 商品控制器：只做参数校验与响应组装，查询逻辑在 goodsService
 */
const goodsService = require('../services/goodsService')
const { success } = require('../utils/response')
const AppError = require('../utils/appError')

/**
 * GET /api/goods
 * 参数：category_id（可选）、keyword（可选）、page（默认1）、page_size（默认10）
 */
exports.list = async (req, res, next) => {
  try {
    const page = parseInt(req.query.page) || 1
    const pageSize = parseInt(req.query.page_size) || 10

    // 参数范围校验
    if (page < 1) throw new AppError('page 必须大于等于 1', 400, 400)
    if (pageSize < 1 || pageSize > 50) throw new AppError('page_size 范围为 1-50', 400, 400)

    const categoryId = parseInt(req.query.category_id) || 0
    const keyword = (req.query.keyword || '').trim()

    const data = await goodsService.listGoods({ categoryId, keyword, page, pageSize })
    success(res, data)
  } catch (err) {
    next(err)
  }
}

/**
 * GET /api/goods/:id
 */
exports.detail = async (req, res, next) => {
  try {
    const id = parseInt(req.params.id)
    if (!id) throw new AppError('商品 id 不合法', 400, 400)

    const data = await goodsService.getGoodsDetail(id)
    success(res, data)
  } catch (err) {
    next(err)
  }
}

/**
 * GET /api/goods/hot
 */
exports.hot = async (req, res, next) => {
  try {
    const list = await goodsService.listHot()
    success(res, list)
  } catch (err) {
    next(err)
  }
}

/**
 * GET /api/goods/new
 */
exports.new = async (req, res, next) => {
  try {
    const list = await goodsService.listNew()
    success(res, list)
  } catch (err) {
    next(err)
  }
}

File: springboot/src/main/java/com/freshmall/controller/GoodsController.java
package com.freshmall.controller;

import com.freshmall.common.AppException;
import com.freshmall.common.Result;
import com.freshmall.service.GoodsService;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

/**
 * 商品接口（对应 Node 版 routes/goods.js）
 * GET /api/goods        分页列表（category_id/keyword/page/page_size）
 * GET /api/goods/hot    热销
 * GET /api/goods/new    新品
 * GET /api/goods/:id    详情
 */
@RestController
@RequestMapping("/api/goods")
@RequiredArgsConstructor
public class GoodsController {

    private final GoodsService goodsService;

    @GetMapping
    public Result<Map<String, Object>> list(
            @RequestParam(defaultValue = "0") int category_id,
            @RequestParam(required = false) String keyword,
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "10") int page_size) {
        if (page < 1) throw new AppException("page 必须大于等于 1", 400, 400);
        if (page_size < 1 || page_size > 50) throw new AppException("page_size 范围为 1-50", 400, 400);
        return Result.success(goodsService.listGoods(category_id, keyword, page, page_size));
    }

    @GetMapping("/hot")
    public Result<List<Map<String, Object>>> hot() {
        return Result.success(goodsService.listHot(10));
    }

    @GetMapping("/new")
    public Result<List<Map<String, Object>>> newGoods() {
        return Result.success(goodsService.listNew(10));
    }

    @GetMapping("/{id}")
    public Result<Map<String, Object>> detail(@PathVariable Integer id) {
        return Result.success(goodsService.getGoodsDetail(id));
    }
}
